lib/fauxpaas/plans/plan.rb

```diff
@@ -0,0 +1,79 @@
+# frozen_string_literal: true
+
+require "fauxpaas/status"
+
+module Fauxpaas
+  module Plans
+
+    # A plan is a series of tasks assembled to achieve a particular purpose. When
+    # a plan is executed, it procedes through the tasks sequentially until either
+    # the series if finished, or a failure is encountered.
+    #
+    # Plans are divided into three phases: prepare, main, and finish, in that order.
+    # These phases are evaluated separately. {#prepare} is called first, and {#main} is
+    # not called until the tasks it defines have each executed successfully. The same
+    # relationship exists between {#main} and {#finish}. This allows for the latter
+    # two phases to determine their tasks at runtime based on changes made by the
+    # previous phases.
+    #
+    # @abstract Plans should be subclasses that override one or more of {#prepare},
+    #   {#main}, and/or {#finish}.
+    class Plan
+
+      # @param target [Artifact,Release] This could be an artifact or a release
```

Comment:
What is this role? We don't send the target any messages here, so what is special about Artifact or Release other than they are sort of what we expect to be passed as we understand things now?

The attr_reader is public, but I don't see it called anywhere. That leads me to believe that we basically don't care what the target is other than indicating the tasks in the plan will be called with it as the single parameter and should be able to operate on it (making this a generic collection/algorithm, in the sense of Array, Enumerable, etc.).